jax/api.py

```diff
@@ -1609,15 +1609,27 @@ def padded_spec(shape_spec):
       return tree_unflatten(out_tree, outs)
   return wrapped_fun
 
+NpDType = Any
 @curry
-def shapecheck(in_shapes, out_shape, fun: Callable):
+def shapecheck(in_shapes, out_shape, fun: Callable,
+               dtypes: Optional[Sequence[NpDType]] = None):
```

Comment:
Perhaps we should name this as `in_dtypes`?

Reply:
I think that this kind of addition to `shapecheck` is important to have, but the actual mechanism does not seem very ergonomic because one has to specify the `in_shapes` and `in_dtypes` separately. Perhaps a better solution is to use a ShapeDtypeStruct where the shape can be a string?

Reply:
There is a way to avoid this problem, by using jax2tf with shape-polymorphic conversion. That will exercise the same code as shapecheck, but will pickup dtype from the actual arguments (like the masking transform). This alternative will exercise more code, e.g., the jax2tf shape-polymorphic conversion, which is why I suggested earlier to just focus on shapecheck. But if shapecheck is blocking us, it is better to go directly to jax2tf.